### PR TITLE
internal/appsec: implement WAF actions for http/grpc

### PR DESCRIPTION
**Pull request by Hellzy**
### What does this PR do?
Add a system of "actions", which are basically callbacks that are executed after the WAF matches specific rules.
Also add a default action - both for grpc and http - that will block a before its execution if said action is called.

### Motivation
This change allows the libraries to react proactively when a rule gets matched by the WAF. The primary example is when
blocking a user depending on their IP. If a blocked IP issues a request, the WAF will match a rule that links to an action that will block the request before it gets a chance of being executed.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Changed code has unit tests for its functionality.
- [ ] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

File: internal/appsec/waf.go [hunk 1]
```diff
@@ -277,6 +297,7 @@ const (
 	serverRequestPathParams           = "server.request.path_params"
 	serverRequestBody                 = "server.request.body"
 	serverResponseStatusAddr          = "server.response.status"
+	httpClientIP                      = "http.client_ip"
```
Comment (Julio-Guerra) [inline internal/appsec/waf.go:300]: Missing `Addr` suffix and the client ip address was badly named in the RFC so I'd suggest to hide it here and follow the same naming convention regardless (no strong opinion on this point):

```suggestion
	serverRequestPathParamsAddr       = "server.request.path_params"
	serverRequestBodyAddr             = "server.request.body"
	serverResponseStatusAddr      = "server.response.status"
	serverRequestClientIPAddr                  = "http.client_ip"
```
(it has formating issues I think)

File: internal/appsec/waf.go [hunk 2]
```diff
@@ -217,7 +237,7 @@ func newGRPCWAFEventListener(handle *waf.Handle, _ []string, timeout time.Durati
 			if md := handlerArgs.Metadata; len(md) > 0 {
 				values[grpcServerRequestMetadata] = md
 			}
-			event := runWAF(wafCtx, values, timeout)
+			event, _ := runWAF(wafCtx, values, timeout)
```
Comment (Julio-Guerra) [inline internal/appsec/waf.go:240]:
```suggestion
			// Run the WAF by ignoring the returned actions, if any, since blocking after the request handler's
			// response is not supported at the moment.
			event, _ := runWAF(wafCtx, values, timeout)
```

File: internal/appsec/waf.go [hunk 3]
```diff
@@ -143,7 +163,7 @@ func newHTTPWAFEventListener(handle *waf.Handle, addresses []string, timeout tim
 					values[serverResponseStatusAddr] = res.Status
 				}
 			}
-			matches := runWAF(wafCtx, values, timeout)
+			matches, _ := runWAF(wafCtx, values, timeout)
```
Comment (Julio-Guerra) [inline internal/appsec/waf.go:166]:
```suggestion
			// Run the WAF by ignoring the returned actions, if any, since blocking after the request handler's
			// response is not supported at the moment.
			matches, _ := runWAF(wafCtx, values, timeout)
```

File: internal/appsec/waf.go [hunk 4]
```diff
@@ -95,9 +95,35 @@ func registerWAF(rules []byte, timeout time.Duration, limiter Limiter, obfCfg *O
...
+			// The WAF event listener got concurrently released
+			return
+		}
+
+		values := map[string]interface{}{}
+		for _, addr := range addresses {
+			if addr == httpClientIP && args.ClientIP.IsValid() {
+				values[httpClientIP] = args.ClientIP.String()
+			}
+		}
```
Comment (Julio-Guerra) [inline internal/appsec/waf.go:113]: The for-loop is not necessary in this simplified ip-blocking-only case:
```suggestion
		if _, exists := addresses[httpClientIP]; exists && args.ClientIP.IsValid() {
			values[httpClientIP] = args.ClientIP.String()
		}
		// TODO: suspicious request blocking by moving here all the addresses available when the request begins
```

File: internal/appsec/waf.go [hunk 5]
```diff
@@ -95,9 +95,35 @@ func registerWAF(rules []byte, timeout time.Duration, limiter Limiter, obfCfg *O
...
+
+		matches, actionIds := runWAF(wafCtx, values, timeout)
+		actionIds = append(actionIds, "block")
```
Comment (Julio-Guerra) [inline internal/appsec/waf.go:116]: Why is this necessary? 🤔

File: internal/appsec/dyngo/instrumentation/common.go [hunk 6]
```diff
@@ -65,6 +65,13 @@ func (s *SecurityEventsHolder) Events() []json.RawMessage {
 	return s.events
 }
 
+// ClearEvents clears the list of stored events
+func (s *SecurityEventsHolder) ClearEvents() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.events = []json.RawMessage{}
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/common.go:72]: Keep the underlying memory as long as you keep the operation:
```suggestion
	s.events = s.events[0:0]
```

File: internal/appsec/waf.go [hunk 7]
```diff
@@ -95,9 +95,35 @@ func registerWAF(rules []byte, timeout time.Duration, limiter Limiter, obfCfg *O
...
+		if len(matches) > 0 {
+			for _, id := range actionIds {
+				actionHandler.Apply(id, op)
+			}
+			op.AddSecurityEvents(matches)
+			log.Debug("appsec: WAF detected an attack before executing the request")
+			if len(actionIds) > 0 {
+				return
+			}
```
Comment (Julio-Guerra) [inline internal/appsec/waf.go:125]: let's make this even more explicit: please add a `blocked` boolean in the operation to set to true when the action was a blocking one and check it here:
```suggestion
			if op.blocked {
				return
			}
```

**Review by Julio-Guerra — CHANGES_REQUESTED** [review]

File: contrib/google.golang.org/grpc/appsec.go [hunk 8]
```diff
@@ -25,6 +25,9 @@ func appsecUnaryHandlerMiddleware(span ddtrace.Span, handler grpc.UnaryHandler)
 	return func(ctx context.Context, req interface{}) (interface{}, error) {
 		md, _ := metadata.FromIncomingContext(ctx)
 		op := grpcsec.StartHandlerOperation(grpcsec.HandlerOperationArgs{Metadata: md}, nil)
+		if op.UnaryHandler != nil {
+			return op.UnaryHandler(ctx, req)
```
Comment (Julio-Guerra) [inline contrib/google.golang.org/grpc/appsec.go:29]:
```suggestion
			return op.UnaryActionHandler(ctx, req)
```

File: contrib/google.golang.org/grpc/appsec.go [hunk 9]
```diff
@@ -44,6 +47,9 @@ func appsecStreamHandlerMiddleware(span ddtrace.Span, handler grpc.StreamHandler
 	return func(srv interface{}, stream grpc.ServerStream) error {
 		md, _ := metadata.FromIncomingContext(stream.Context())
 		op := grpcsec.StartHandlerOperation(grpcsec.HandlerOperationArgs{Metadata: md}, nil)
+		if op.StreamHandler != nil {
+			return op.StreamHandler(srv, stream)
```
Comment (Julio-Guerra) [inline contrib/google.golang.org/grpc/appsec.go:51]:
```suggestion
			return op.StreamActionHandler(srv, stream)
```

File: internal/appsec/dyngo/instrumentation/grpcsec/grpc.go [hunk 10]
```diff
@@ -39,6 +41,8 @@ type (
 		dyngo.Operation
 		instrumentation.TagsHolder
 		instrumentation.SecurityEventsHolder
+		UnaryHandler  grpc.UnaryHandler
+		StreamHandler grpc.StreamHandler
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/grpcsec/grpc.go:45]: what about making it super simple for now by simply having a `blocked code.Code` field here in place of those two?
We know the current set of "more dynamic action parameters" don't apply to gRPC for now.

File: internal/appsec/dyngo/instrumentation/httpsec/actions.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+// BlockRequestAction is the parameter struct used to perform actions of kind ActionBlockRequest
+type BlockRequestAction struct {
+	Action
+	// Status is the return code to use when blocking the request
+	Status int
+	// Template is the payload template to use to write the response (html or json)
+	Template string
+	// handler is the http handler be used to block the request (see wrap())
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/actions.go:24]:
```suggestion
	// handler is the http handler to use to block the request (see wrap())
```

Comment (Julio-Guerra) [reply internal/appsec/dyngo/instrumentation/httpsec/actions.go:24]: `wrap()` doesn't exist, right?

File: internal/appsec/dyngo/instrumentation/httpsec/actions.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+// Currently, only the default "block" action is supported
+func NewActionsHandler() *ActionsHandler {
+	handler := ActionsHandler{
+		actions: map[string]Action{},
+	}
+	// Register the default "block" action as specified in the RFC for HTTP blocking
+	handler.RegisterAction("block", &BlockRequestAction{
+		Status:   403,
+		Template: "html",
+	})
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/actions.go:53]: The spec says the default blocking action is:

> The default block action is defined as follow:
id: block
type: block_request
parameters:
   status_code: 403
   type: auto
   grpc_status_code: 10

You don't seem to follow it here.

File: internal/appsec/dyngo/instrumentation/httpsec/actions.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+// Action is used to identify any action kind
+type Action interface {
+	isAction()
+}
+
+// BlockRequestAction is the parameter struct used to perform actions of kind ActionBlockRequest
+type BlockRequestAction struct {
+	Action
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/actions.go:19]:
```suggestion
	Action // to implement the interface
```

Comment (Julio-Guerra) [reply internal/appsec/dyngo/instrumentation/httpsec/actions.go:19]: BTW, I was thinking of what are the pros and cons of this "hack" and not that you are impacting the sizeof of your struct (IIRC, an interface is made of 2 pointers), while implementing the method doesn't impact the size of the struct ;-)

File: internal/appsec/dyngo/instrumentation/httpsec/actions.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+	handler := ActionsHandler{
+		actions: map[string]Action{},
+	}
+	// Register the default "block" action as specified in the RFC for HTTP blocking
+	handler.RegisterAction("block", &BlockRequestAction{
+		Status:   403,
+		Template: "html",
+	})
+
+	return &handler
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/actions.go:55]: Bonus but worth knowing and understanding:
this type only wraps a map alone, so returning `&handler` isn't worth it and you should return `handler` instead to save an heap allocation here (it doesn't mean the value escapes on the heap somewhere else).

File: internal/appsec/dyngo/instrumentation/httpsec/http.go [hunk 15]
```diff
@@ -72,25 +75,44 @@ func MonitorParsedBody(ctx context.Context, body interface{}) {
...
+			op.ClearActions()
+		}
 		defer func() {
 			var status int
 			if mw, ok := w.(interface{ Status() int }); ok {
 				status = mw.Status()
 			}
 
-			events := op.Finish(HandlerOperationRes{Status: status})
+			events = append(events, op.Finish(HandlerOperationRes{Status: status})...)
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/http.go:109]: We could also not `ClearEvents()` and get the full list of events at once here without this append

File: internal/appsec/dyngo/instrumentation/httpsec/actions.go (hunk 11 above)
Comment (Hellzy) [reply internal/appsec/dyngo/instrumentation/httpsec/actions.go:24]: Not anymore, nice catch. Edited you suggestion to remove that part too.

File: internal/appsec/dyngo/instrumentation/httpsec/actions.go (hunk 13 above)
Comment (Hellzy) [reply internal/appsec/dyngo/instrumentation/httpsec/actions.go:19]: Let's stick to implementing only then :+1: 
```suggestion
```

File: internal/appsec/waf.go (hunk 4 above)
Comment (Hellzy) [reply internal/appsec/waf.go]: I was to quick to commit this. Actually `addresses` is a slice here, not a map

File: internal/appsec/waf.go (hunk 7 above)
Comment (Hellzy) [reply internal/appsec/waf.go:123]: The operation hasn't necessarily been blocked yet though, we only know that the action is supposed to block it. I think it would be semantically more correct to check the action type, or maybe `ActionHandler.Apply()` could return some status indicating whether applying this action should interrupt the flow here. WDYT?

File: internal/appsec/dyngo/instrumentation/httpsec/actions.go [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+// the action will not be registered
+func (h *ActionsHandler) RegisterAction(id string, action Action) {
+	switch a := action.(type) {
+	case *BlockRequestAction:
+		payload := blockedPayload(a)
+		a.handler = http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+			writer.WriteHeader(a.Status)
+			writer.Write(payload)
+		})
+		h.actions[id] = a
```
[the remark this answers is not in the document]
Comment (Hellzy) [reply internal/appsec/dyngo/instrumentation/httpsec/actions.go]: https://github.com/DataDog/dd-trace-go/pull/1533/commits/9f24219a3f1a4480541a4aa18f3b05ed108d2929

File: internal/appsec/waf.go (hunk 7 above)
Comment (Hellzy) [reply internal/appsec/waf.go]: See https://github.com/DataDog/dd-trace-go/pull/1533/commits/784323059db23d711f065ed523f853a3fde57ebb

File: internal/appsec/dyngo/instrumentation/httpsec/http.go (hunk 15 above)
Comment (Hellzy) [reply internal/appsec/dyngo/instrumentation/httpsec/http.go]: https://github.com/DataDog/dd-trace-go/pull/1533/commits/784323059db23d711f065ed523f853a3fde57ebb

File: internal/appsec/dyngo/instrumentation/grpcsec/grpc.go (hunk 10 above)
Comment (Hellzy) [reply internal/appsec/dyngo/instrumentation/grpcsec/grpc.go]: Took a stab at it in 9f6411eafd0a2

File: internal/appsec/dyngo/instrumentation/httpsec/http.go [hunk 17]
```diff
@@ -72,25 +75,44 @@ func MonitorParsedBody(ctx context.Context, body interface{}) {
 // HandlerOperationRes.
 func WrapHandler(handler http.Handler, span ddtrace.Span, pathParams map[string]string) http.Handler {
 	instrumentation.SetAppSecEnabledTags(span)
+	applyActions := func(op *Operation) {
```
[the remark this answers is not in the document]
Comment (Hellzy) [reply internal/appsec/dyngo/instrumentation/httpsec/http.go]: See 49a7c278912

File: internal/appsec/dyngo/instrumentation/httpsec/http.go [hunk 18]
```diff
@@ -261,3 +309,64 @@ func (OnSDKBodyOperationFinish) ListenedType() reflect.Type { return sdkBodyOper
...
...ground: radial-gradient(circle at 26% 19%, #fff, #f4f7f9); display: -webkit-box; display: -ms-flexbox; display: flex; -webkit-box-pack: center; -ms-flex-pack: center; justify-content: center; -webkit-box-align: center; -ms-flex-align: center; align-items: center; -ms-flex-line-pack: center; align-content: center; width: 100%; min-height: 100vh; line-height: 1; flex-direction: column}p{display: block}main{text-align: center; flex: 1; display: -webkit-box; display: -ms-flexbox; display: flex; -webkit-box-pack: center; -ms-flex-pack: center; justify-content: center; -webkit-box-align: center; -ms-flex-align: center; align-items: center; -ms-flex-line-pack: center; align-content: center; flex-direction: column}p{font-size: 18px; line-height: normal; color: #646464; font-family: sans-serif; font-weight: 400}a{color: #4842b7}footer{width: 100%; text-align: center}footer p{font-size: 16px}</style></head><body> <main> <p>Sorry, you cannot access this page. Please contact the customer service team.</p></main> <footer> <p>Security provided by <a href="https://www.datadoghq.com/product/security-platform/application-security-monitoring/" target="_blank">Datadog</a></p></footer></body></html>`)
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/http.go:353]: Bonus: use `//go:embed`

File: internal/appsec/waf.go [hunk 19]
```diff
@@ -188,6 +212,34 @@ func newGRPCWAFEventListener(handle *waf.Handle, _ []string, timeout time.Durati
...
+		// The same address is used for gRPC and http when it comes to client ip
+		values := map[string]interface{}{}
+		for _, addr := range addresses {
+			if addr == httpClientIPAddr && handlerArgs.ClientIP.IsValid() {
+				values[httpClientIPAddr] = handlerArgs.ClientIP.String()
+			}
+		}
+
+		matches, actionIds := runWAF(wafCtx, values, timeout)
+		wafCtx.Close()
```
Comment (Julio-Guerra) [inline internal/appsec/waf.go:230]: can you please defer this call right after line 219? This avoids leaking memory in case any of the lines between 219 and this one panic in loop on each request.

File: contrib/google.golang.org/grpc/appsec.go [hunk 20]
```diff
@@ -12,19 +12,22 @@ import (
...
+	"google.golang.org/grpc/status"
 )
 
 // UnaryHandler wrapper to use when AppSec is enabled to monitor its execution.
 func appsecUnaryHandlerMiddleware(span ddtrace.Span, handler grpc.UnaryHandler) grpc.UnaryHandler {
 	instrumentation.SetAppSecEnabledTags(span)
 	return func(ctx context.Context, req interface{}) (interface{}, error) {
 		md, _ := metadata.FromIncomingContext(ctx)
-		op := grpcsec.StartHandlerOperation(grpcsec.HandlerOperationArgs{Metadata: md}, nil)
+		ip := httpsec.IPFromHeaders(httpsec.NormalizeHTTPHeaders(md), "")
```
Comment (Julio-Guerra) [inline contrib/google.golang.org/grpc/appsec.go:29]: If you add a unit test for grpc's ip-blocking, it shouldn't be blocked because `NormalizeHTTPHeaders` is filtering-out all the headers but the collected ones (in the span tags). This function actually badly named because it doesn't reflect the fact they get filtered too :-/

**Review by knusbaum — DISMISSED** [review]
All good from trace side.

**Review by Julio-Guerra — APPROVED** [review]

**Review by ajgajg1134 — APPROVED** [review]
👍 for APM
